File: module_B_dl/task_1_image_recognition/image_check.py
#!/usr/bin/env python3

import sys
from pathlib import Path


DEFAULT_IMAGE = Path("/home/pgq/JS/RobotLearning/module_B_dl/task_1_image_recognition/images/action.png")
BLESSING_LINES = [
    "恭喜你成功识别到这张祝福图片。",
    "愿新的训练任务顺利完成，代码一路通过。",
]


def detect_image_type(header: bytes) -> str:
    if header.startswith(b"\x89PNG\r\n\x1a\n"):
        return "PNG image"
    if header.startswith(b"\xff\xd8\xff"):
        return "JPEG image"
    if header.startswith((b"GIF87a", b"GIF89a")):
        return "GIF image"
    if header.startswith(b"BM"):
        return "BMP image"
    if len(header) >= 12 and header[:4] == b"RIFF" and header[8:12] == b"WEBP":
        return "WEBP image"
    return "Unknown image format"


def main() -> int:
    image_path = Path(sys.argv[1]) if len(sys.argv) > 1 else DEFAULT_IMAGE

    if not image_path.is_file():
        print(f"Error: image file not found: {image_path}")
        return 1

    with image_path.open("rb") as image_file:
        header = image_file.read(16)

    image_type = detect_image_type(header)
    file_size = image_path.stat().st_size

    print("===== Image Check Result =====")
    print(f"Image path : {image_path}")
    print(f"File size  : {file_size} bytes")
    print(f"Image type : {image_type}")
    print("Message    :")
    for line in BLESSING_LINES:
        print(f"  {line}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: module_B_dl/task_1_image_recognition/image_check.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

const std::string DEFAULT_IMAGE = "/home/unitree/training_camp/module_B_dl/";
const std::vector<std::string> BLESSING_LINES = {
    "恭喜你成功识别到这张祝福图片。",
    "愿新的训练任务顺利完成，代码一路通过。"
};

bool startsWith(const std::vector<unsigned char>& data,
                const std::vector<unsigned char>& pattern) {
    if (data.size() < pattern.size()) {
        return false;
    }

    for (std::size_t i = 0; i < pattern.size(); ++i) {
        if (data[i] != pattern[i]) {
            return false;
        }
    }

    return true;
}

std::string detectImageType(const std::vector<unsigned char>& header) {
    if (startsWith(header, {0x89, 'P', 'N', 'G', 0x0D, 0x0A, 0x1A, 0x0A})) {
        return "PNG image";
    }
    if (startsWith(header, {0xFF, 0xD8, 0xFF})) {
        return "JPEG image";
    }
    if (startsWith(header, {'G', 'I', 'F', '8', '7', 'a'}) ||
        startsWith(header, {'G', 'I', 'F', '8', '9', 'a'})) {
        return "GIF image";
    }
    if (startsWith(header, {'B', 'M'})) {
        return "BMP image";
    }
    if (header.size() >= 12 &&
        header[0] == 'R' && header[1] == 'I' && header[2] == 'F' && header[3] == 'F' &&
        header[8] == 'W' && header[9] == 'E' && header[10] == 'B' && header[11] == 'P') {
        return "WEBP image";
    }

    return "Unknown image format";
}

int main(int argc, char* argv[]) {
    std::string imagePath = argc > 1 ? argv[1] : DEFAULT_IMAGE;

    std::ifstream imageFile(imagePath, ____);
    if (!imageFile) {
        std::cerr << "Error: image file not found: " << imagePath << std::endl;
        return 1;
    }

    std::vector<unsigned char> header(16);
    imageFile.read(reinterpret_cast<char*>(header.data()), header.size());
    header.resize(static_cast<std::size_t>(imageFile.gcount()));

    imageFile.clear();
    imageFile.seekg(0, std::ios::end);
    std::streamoff fileSize = imageFile.tellg();

    std::string imageType = ____(header);

    std::cout << "===== Image Check Result =====" << std::endl;
    std::cout << "Image path : " << imagePath << std::endl;
    std::cout << "File size  : " << fileSize << " bytes" << std::endl;
    std::cout << "Image type : " << imageType << std::endl;
    std::cout << "Message    :" << std::endl;
    for (const std::string& line : BLESSING_LINES) {
        std::cout << "  " << line << std::endl;
    }

    return 0;
}
